alchemist/alchemist-fxui/src/main/kotlin/it/unibo/alchemist/boundary/monitors/AbstractFXDisplay.kt

```diff
@@ -0,0 +1,287 @@
+import java.util.stream.Collectors
+import java.util.stream.Stream
+
+/**
+ * Base abstract class for each display able to graphically represent a 2D space and simulation.
+ *
+ * @param <T> The type which describes the [Concentration] of a molecule
+ * @param <P> The type of position
+ */
+@SuppressFBWarnings(
+    "NP_NONNULL_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR",
+    "Field is initialized in the initialize function"
+)
+abstract class AbstractFXDisplay<T, P : Position2D<P>>
+/**
+ * Main constructor. It lets the developer specify the number of steps.
+ *
+ * @param steps the number of steps
+ * @see .setStep
+ */
+@JvmOverloads constructor(steps: Int = DEFAULT_NUMBER_OF_STEPS) : Pane(), FXOutputMonitor<T, P> {
+
+    private val effectStack: ObservableList<EffectGroup<P>> = FXCollections.observableArrayList()
+    private val mutex = Semaphore(1)
+    private val mayRender = AtomicBoolean(true)
+    private var step: Int = 0
+    @Volatile private var firstTime: Boolean = false
+    private var realTime: Boolean = false
+    @Volatile private var commandQueue: ConcurrentLinkedQueue<() -> Unit> = ConcurrentLinkedQueue()
+    private var viewStatus = DEFAULT_VIEW_STATUS
+    private lateinit var wormhole: BidimensionalWormhole<P>
+    private lateinit var zoomManager: ZoomManager
+    private val interactions: InteractionManager<T, P> by lazy { InteractionManager(this) }
+    private val effectsCanvas = Canvas()
+    /**
+     * Group dedicated for painting the background.
+     */
+    val background = Group()
+
+    init {
+        firstTime = true
+        setStep(steps)
+        val repaintOnResize = ChangeListener<Number> { _, _, _ ->
+            repaint()
+        }
+        widthProperty().addListener(repaintOnResize)
+        heightProperty().addListener(repaintOnResize)
+        effectsCanvas.isMouseTransparent = true
+        effectsCanvas.widthProperty().bind(widthProperty())
+        effectsCanvas.heightProperty().bind(heightProperty())
+        children.addAll(background, effectsCanvas, interactions.canvases)
+    }
+
+    override fun getViewStatus(): FXOutputMonitor.ViewStatus {
+        return this.viewStatus
+    }
+
+    override fun setViewStatus(viewStatus: FXOutputMonitor.ViewStatus) {
+        this.viewStatus = viewStatus
+    }
+
+    override fun getStep(): Int {
+        return this.step
+    }
+
+    /**
+     * {@inheritDoc}.
+     *
+     * @throws IllegalArgumentException if the step is not bigger than 0
+     */
+    @Throws(IllegalArgumentException::class)
+    override fun setStep(step: Int) {
+        if (step <= 0) {
+            throw IllegalArgumentException("The parameter must be a positive integer")
+        }
+        this.step = step
+    }
+
+    override fun isRealTime(): Boolean {
+        return this.realTime
+    }
+
+    override fun setRealTime(realTime: Boolean) {
+        this.realTime = realTime
+    }
+
+    override fun repaint() {
+        mutex.acquireUninterruptibly()
```

Comment:
Why do you need a mutex?

Reply:
`Repaint` is called by both the background thread as well as the JavaFX thread, so I believe this is necessary in order to guarantee that the different `repaint` calls are done sequentially and never in parallel.

Reply:
Uhmm... this sounds strange: shouldn't be there a single JFX thread allowed to run `repaint`? What is the "background thread"?

Reply:
When a simulation step is done, the simulation thread (what I previously called the background thread) calls `stepDone` which also calls `repaint`. But it is possible for the JavaFX thread to call `repaint` as well, for instance through a ChangeListener that is added to width and height properties so that the effects are repainted when the monitor's dimensions change.